### PR TITLE
GG-39005 Added auto shutdown for maintenance mode

### PR DESCRIPTION
**Pull request by Erixonich**
https://ggsystems.atlassian.net/browse/GG-39005

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java [hunk 1]
```diff
@@ -231,6 +243,14 @@ else if (isMaintenanceMode()) {
                     "Restart the node to get it back to normal operations.");
             }
         }
+
+        if (isMaintenanceMode() && autoShutdown) {
+            if (log.isInfoEnabled())
+                log.info("Starting thread which will check active maintenance tasks. " +
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java:249]: I would rephrase it into something like: 
`Starting a thread that will shut down the node when all maintenance tasks are completed.`

File: modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java [hunk 2]
```diff
@@ -344,4 +364,24 @@ private void proceedWithMaintenance() {
     @Nullable @Override public MaintenanceTask requestedTask(String maintenanceTaskName) {
         return requestedTasks.get(maintenanceTaskName);
     }
+
+    /** Starts a thread which will shut down the node when all maintenance tasks would be completed */
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java:368]:
```suggestion
    /** Starts a thread which will shut down the node when all maintenance tasks would be completed. */
```

File: modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java [hunk 3]
```diff
@@ -344,4 +364,24 @@ private void proceedWithMaintenance() {
...
+
+        IgniteThread t = new IgniteThread(new GridWorker(igniteInstanceName, "maintenance-shutdown-thread", log) {
+            @Override protected void body() {
+                while (true) {
+                    if (activeTasks.isEmpty() && Ignition.state(igniteInstanceName) == IgniteState.STARTED) {
+                        G.stop(igniteInstanceName, false);
+                        return;
+                    }
+
+                    LockSupport.parkNanos(100_000_000);
```
Comment (tkalkirill) [inline modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java:380]: Why are you using `LockSupport#parkNanos` instead of `Thread#sleep`?
Maybe periodically log the number of unfinished tasks?
Can we log when the tasks are over and we go to shut down the node?

File: modules/core/src/test/java/org/apache/ignite/testsuites/IgnitePdsTestSuite8.java [hunk 4]
```diff
@@ -21,16 +21,7 @@
...
-import org.apache.ignite.internal.processors.cache.persistence.IgnitePdsConsistencyOnDelayedPartitionOwning;
-import org.apache.ignite.internal.processors.cache.persistence.IgnitePdsDefragmentationEncryptionTest;
-import org.apache.ignite.internal.processors.cache.persistence.IgnitePdsDefragmentationRandomLruEvictionTest;
-import org.apache.ignite.internal.processors.cache.persistence.IgnitePdsDefragmentationTest;
-import org.apache.ignite.internal.processors.cache.persistence.IgnitePdsRecoveryAfterFileCorruptionTest;
-import org.apache.ignite.internal.processors.cache.persistence.MaintenanceClearCacheFolderTest;
-import org.apache.ignite.internal.processors.cache.persistence.MaintenancePersistenceTaskTest;
-import org.apache.ignite.internal.processors.cache.persistence.PagesPossibleCorruptionDiagnosticTest;
-import org.apache.ignite.internal.processors.cache.persistence.PendingTreeCorruptionTest;
+import org.apache.ignite.internal.processors.cache.persistence.*;
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/testsuites/IgnitePdsTestSuite8.java:24]: https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines#CodingGuidelines-PackageImporting

Please **Use per-class importing.**

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        consistentId = cfg.getConsistentId().toString();
+
+        return cfg;
+    }
+
+    /** Node will auto shut down once all maintenance tasks are completed. */
+    @Test
+    @WithSystemProperty(key = "MM_AUTO_SHUTDOWN_AFTER_RECOVERY", value = "true")
+    public void testAutoShutdown() throws Exception {
+        IgniteEx ign0 = startGrid(0);
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java:68]: I don’t quite understand what exactly 2 nodes are for? Is one not enough for the test?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        consistentId = cfg.getConsistentId().toString();
+
+        return cfg;
+    }
+
+    /** Node will auto shut down once all maintenance tasks are completed. */
+    @Test
+    @WithSystemProperty(key = "MM_AUTO_SHUTDOWN_AFTER_RECOVERY", value = "true")
+    public void testAutoShutdown() throws Exception {
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java:67]: Let's add a test where the sys_prop is false.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java (hunk 5 above)
Comment (Erixonich) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java:68]: Second node is required to check that first left cluster properly.

File: modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java (hunk 3 above)
Comment (Erixonich) [reply modules/core/src/main/java/org/apache/ignite/internal/maintenance/MaintenanceProcessor.java:380]: done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java (hunk 5 above)
Comment (Erixonich) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java:68]: That's was incorrect

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java (hunk 6 above)
Comment (Erixonich) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/MaintenanceAutoRecoveryTest.java:67]: ok

**Review by ibessonov — APPROVED** [review]
